lighthouse-core/audits/seo/tap-targets.js

```diff
@@ -0,0 +1,240 @@
+  // score (had negative scores before!)
+
+  const failures = [];
+
+  function hasFailure(targetA, targetB) {
+    return failures.some(failure => {
+      return failure.targetA === targetA && failure.targetB === targetB;
+    });
+  }
+
+  for (let i = 0; i < count; i++) {
+    for (let j = 0; j < count; j++) {
+      if (i === j) {
+        continue;
+      }
+
+      const targetA = targets[i];
+      const targetB = targets[j];
+
+      if (hasFailure(targetB, targetA)) {
+        // console.log('already has failure', {i, j});
+        continue;
+      }
+
+      let maxOverlap = 0;
+
+      targetA.bcrs.forEach(bcrA => {
+        const fingerAtCenter = getFingerAtCenter(bcrA);
+        bcrA.fingerAtCenter = fingerAtCenter;
+
+        targetB.bcrs.forEach(bcrB => {
+          let overlap = getOverlapDistanceBetweenRectangles(
+            fingerAtCenter,
+            bcrB
+          );
+
+          // console.log(fingerAtCenter, bcrB);
+
+          if (overlap > maxOverlap) {
+            maxOverlap = overlap;
+          }
+
+          overlap = getOverlapDistanceBetweenRectangles(
+            fingerAtCenter,
+            getFingerAtCenter(bcrB)
+          );
+          if (overlap > maxOverlap) {
+            maxOverlap = overlap;
+          }
+        });
+      });
+
+      // console.log({maxOverlap, a: targetA.snippet, b: targetB.snippet});
+
+      if (maxOverlap > 0) {
+        failures.push({
+          targetA,
+          targetB,
+          overlap: maxOverlap,
+        });
+      }
+    }
+  }
+
+  // console.log({
+  //   failingTapTargetsCount: tapTargetsWithFailures.size,
+  //   failureCount: failures.length,
+  // });
+
+  return failures;
+}
+
+class TapTargets extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'tap-targets',
+      title: 'Tap targets are sized appropriately',
+      failureTitle: 'Tap targets are not sized appropriately',
+      description:
+        'Interactive elements like buttons and links should be large enough, and have enough space around them, to be easy enough to tap without overlapping onto other elements. [Learn more](https://developers.google.com/web/fundamentals/accessibility/accessible-styles#multi-device_responsive_design).',
+      requiredArtifacts: ['Viewport', 'TapTargets'],
+    };
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @return {LH.Audit.Product}
+   */
+  static audit(artifacts) {
+    const hasViewportSet = ViewportAudit.audit(artifacts).rawValue;
+    if (!hasViewportSet) {
+      return {
+        rawValue: false,
+        explanation:
+          'Tap targets are too small because of a missing viewport config',
+      };
+    }
+
+    // console.time('tooCloseTargets');
+    const tooClose = tooCloseTargets(artifacts.TapTargets);
+
+
+    // console.timeEnd('tooCloseTargets');
+    /** @type {Array<{node: LH.Audit.DetailsRendererNodeDetailsJSON, issue: string}>} */
+    const failures = [];
```

Comment:
so these are each tap target pairs where `targetA` is unique across all `failures` and `targetB` is the target with the most overlap with that `targetA`, correct?

I wonder if it might be a little too confusing to the user given the framing focused on tap targets being too small. WDYT about listing the nodes that failed with a filter on node total area? We could then list the selectors of the elements that are too close to them.

Reply:
> so these are each tap target pairs where targetA is unique across all failures and targetB is the target with the most overlap with that targetA, correct?

We report multiple failures if there are multiple overlapping targets, rather than just reporting the biggest overlap. (The logic we have for getting the largest overlap is for different client rects.)

> I wonder if it might be a little too confusing to the user given the framing focused on tap targets being too small. WDYT about listing the nodes that failed with a filter on node total area? We could then list the selectors of the elements that are too close to them.

Good point about it being confusing, I'll talk to Rick about the framing and how we show the results.

Are you suggesting we should surface the node area in the results? Or just filter out failures that have a large tap target?

If a tap target is bigger than 36x36px then it won't fail. (We suggest 48x48px but allow some overlap.)

Reply:
> If a tap target is bigger than 36x36px then it won't fail. (We suggest 48x48px but allow some overlap.)

Right, but what I'm trying to get at is we're listing pairs, not individual elements. Maybe I misunderstood but it seems like it's possible for either targetA or targetB in the list to be large and the other one to be the target that's too small. I'm trying to see if we can reframe the results a bit so that each row is a tap target that is too small, and then one of the columns can list the elements that are overlapping it.

Reply:
`targetA` is always too small. If you try tapping on `targetA` you might accidentally tap `targetB`.

![screen shot 2018-11-08 at 15 37 48](https://user-images.githubusercontent.com/1303660/48209124-49dd1980-e36c-11e8-8bd4-de18ea9de316.png)

(A tap target overlapping another target's finger is highlighted by the orange outline.)

> then one of the columns can list all the elements that are overlapping it.

Yeah that's possible, but might be harder to format with the "extra space needed" info. 95% of the time there's only one overlapping tap target.

Reply:
Ah gotcha, OK well I'll defer to @rviscomi then just wanted to bring it up to make sure we don't get too impl focused compared to what we told users we'd be showing them :)

Reply:
Update after talking to @rviscomi:

- We'll tell people tap targets should be 48x48px
- All tap targets that are too small* are shown in the table
- Score is only impacted if there is an overlapping element nearby
- We don't show "extra space needed"

So columns are: Too Small Element | Size | Overlapping Element

* probably based on the of the largest merged area if we have multiple BCRs